File: 7-reverse-integer/reverse_integer.c
#include <stdio.h>
#include <math.h>

int reverse(int x) {
    int upper_limit = (pow(2, 31) - 1) / 10;
    int lower_limit = -(pow(2, 31) / 10);
    int ans = 0;

    while (x) {
        if (ans < lower_limit || ans > upper_limit) {
            return 0;
        }

        ans = ans * 10 + x % 10;
        x /= 10;
    }

    return ans;
}

File: 7-reverse-integer/reverse_integer.rb
UPPER_LIMIT = (2**31 - 1) / 10
LOWER_LIMIT = -(2**31 / 10)

# @param {Integer} x
# @return {Integer}
def reverse(x)
  ans = 0
  n = x.abs

  while n.positive?
    return 0 if ans < LOWER_LIMIT || ans > UPPER_LIMIT

    ans = ans * 10 + n % 10
    n /= 10
  end

  x.negative? ? -ans : ans
end
